accepted/0042-isolated-mode.md

```diff
@@ -0,0 +1,347 @@
+- _Recommended by Node.js_, this approach is actually the recommended approach by [the Node.js documentation](https://nodejs.org/api/modules.html#modules_addenda_package_manager_tips).
+
+# Implementation
+
+## Simple Explanation
+
+The packages are laid out on disk with a flat structure in a folder called the "package store". The dependencies between packages will be expressed by creating symlinks between the various packages.
+
+## How does it work?
+
+This strategy is based on the fact that [Node.js module resolution algorithm](https://nodejs.org/api/modules.html#modules_all_together) follows symlinks and works with them exactly the same as if they were real folders.
+
+Additionally, once a module is resolved, the resolution algorithm calls 'realpath()' on the result. This means that the resolution algorithm always returns a real path. This allows to setup an arbitrary complex dependency graph while making sure Node.js does not create more than one instance of a given module.
+
+## Detailed Explanation
+
+### Structure
+
+In the workspaces:
+
+- Each workspace has its own `node_modules` folder.
+- Each `node_modules` folder contain symlinks.
+- The symlinks in a `node_modules` folder correspond to the matching workspace's dependencies.
+- Each symlink is given the same name as the dependency it represents.
+- Each symlink points to a location in the package store where the corresponding package is installed.
+
+In the package store:
+
+- The package store is a folder stored in the root/project level `node_modules` folder.
+- The package store is named `.npm`.
+- The package store contains folders, one for each dependency installed. These dependencies can be direct workspace dependencies or transitive dependencies.
+- These folders have a name containing the name of the package, its version and a hash of its content plus the hash of its resolved peer-dependencies.
+- These folders contain a `node_modules` folder.
+- These `node_modules` folders contain multiple folders, one for each dependency of that package and one for the package itself.
+- All these folders are named based on their respective package names.
+- The folder representing the package itself contains the code of that package. 
+- The folders representation the dependencies of the package are symlinks to the corresponding package in the store.
+- The dependencies of each packages are installed exactly the same way as workspaces' dependencies are installed.
+
+This structure may seem more complicated than necessary. The reasons for this complexity are:
+
+- to make it possible for a package to import itself using its own name. For example, the following code `require.resolve("foo");` in `node_modules/.npm/foo@1.0.0-1234/node_modules/foo/bar.js` will correctly return `node_modules/.npm/foo@1.0.0-1234/node_modules/foo/index.js`.
+- to prevent creating circular symlinks, even when packages have circular dependencies.
+
+#### Peer dependencies
```

Comment:
I think we can probably avoid a lot of this peerDep/edge case analysis if we specify that we are going to implement the isolation mode as a post-buildIdealTree conversion step.

That is, even though there are dependency graph shapes that may be enabled by isolated-mode, which are not enabled by hoisted mode, some some of these may be desirable, we are going to instead favor interoperability and reducing the degree of change in the dependency contracts enforced by npm.

So, whereas hoisted mode does the following:

1. Calculate hoisted ideal tree to satisfy dependency graph.
2. Reify ideal tree.
3. Save ideal tree to package-lock.json

instead, npm will:

1. Calculate hoisted ideal tree to satisfy dependency graph.
2. Capture hoisted tree lockfile metadata.
3. Convert hoisted tree into isolated tree.
4. Reify isolated tree.
5. Save hoisted tree lockfile metadata to package-lock.json

Thus, any situation not currently resolvable by hoisted-mode will not be possible with isolated-mode.  However, this ensures that any project may be reified using _either_ hoisted or isolated mode, or switched between them, without any changes to the project lockfile.

Anything that causes an `ERESOLVE` error in hoisted mode will also cause an `ERESOLVE` error in isolated mode.

Then the task is to document how that conversion process is accomplished.  Imo, that ought to end up in this RFC document, even if it happens after some implementation work has been done (which is probably a good idea, since we're bound to find some edge cases we haven't thought of!)